da/celestia/celestia.go

```diff
@@ -285,10 +357,264 @@ func (c *DataAvailabilityLayerClient) RetrieveBatches(dataLayerHeight uint64) da
+
+	dah, err := c.getDataAvailabilityHeaders(height)
+	if err != nil {
+		return nil, nil, err
+	}
+	rowProof, err := c.getRowProof(height, dah, blob, commitment, proof)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return proof, rowProof, nil
+}
+
+// Submit submits the Blobs to Data Availability layer.
+func (c *DataAvailabilityLayerClient) submit(daBlobs []da.Blob) (uint64, []da.Commitment, error) {
+	blobs, commitments, err := c.blobsAndCommitments(daBlobs)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	options := openrpc.DefaultSubmitOptions()
+
+	if c.config.GasPrices >= 0 {
+		blobSizes := make([]uint32, len(blobs))
+		for i, blob := range blobs {
+			blobSizes[i] = uint32(len(blob.Data))
+		}
+
+		estimatedGas := EstimateGas(blobSizes, DefaultGasPerBlobByte, DefaultTxSizeCostPerByte)
+		gasWanted := uint64(float64(estimatedGas) * c.config.GasAdjustment)
+		fees := c.calculateFees(gasWanted)
+		options.Fee = fees
+		options.GasLimit = gasWanted
+	}
+	ctx, cancel := context.WithTimeout(c.ctx, c.txPollingRetryDelay)
+	defer cancel()
+
+	height, err := c.rpc.Submit(ctx, blobs, options)
+
+	if err != nil {
+		return 0, nil, err
+	}
+	c.logger.Info("Successfully submitted blobs to Celestia", "height", height, "gas", options.GasLimit, "fee", options.Fee)
+
+	return height, commitments, nil
+}
+
+func (c *DataAvailabilityLayerClient) getProof(height uint64, commitment da.Commitment) (*blob.Proof, error) {
+
+	c.logger.Info("Getting proof via RPC call")
+	ctx, cancel := context.WithTimeout(c.ctx, c.txPollingRetryDelay)
+	defer cancel()
+
+	proof, err := c.rpc.GetProof(ctx, height, c.config.NamespaceID.Bytes(), commitment)
+	if err != nil {
+		return nil, err
+	}
+
+	return proof, nil
+
+}
+
+func (c *DataAvailabilityLayerClient) blobsAndCommitments(daBlobs []da.Blob) ([]*blob.Blob, []da.Commitment, error) {
+	var blobs []*blob.Blob
+	var commitments []da.Commitment
+	for _, daBlob := range daBlobs {
+		b, err := blob.NewBlobV0(c.config.NamespaceID.Bytes(), daBlob)
+		if err != nil {
+			return nil, nil, err
+		}
+		blobs = append(blobs, b)
+
+		commitment, err := blob.CreateCommitment(b)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		commitments = append(commitments, commitment)
+	}
+	return blobs, commitments, nil
+}
+
+func (c *DataAvailabilityLayerClient) validateProof(height uint64, commitment da.Commitment, proof *blob.Proof) (bool, error) {
+
+	c.logger.Info("Getting inclusion validation via RPC call")
+	ctx, cancel := context.WithTimeout(c.ctx, c.txPollingRetryDelay)
+	defer cancel()
+	isIncluded, error := c.rpc.Included(ctx, height, c.config.NamespaceID.Bytes(), proof, commitment)
+	return isIncluded, error
+}
+
+func (c *DataAvailabilityLayerClient) getDataAvailabilityHeaders(height uint64) (*header.DataAvailabilityHeader, error) {
+
+	c.logger.Info("Getting Celestia extended headers via RPC call")
+	ctx, cancel := context.WithTimeout(c.ctx, c.txPollingRetryDelay)
+	defer cancel()
```

Comment:
no need for it

Reply:
it's there to force  a timeout in case rpc call does not timeout by itself (which was happening for some problems in celestia)